executor/builder.go

```diff
@@ -287,24 +293,54 @@ func (b *executorBuilder) buildIndexScan(v *plan.IndexScan) Executor {
 
 	e.Ranges = make([]*IndexRangeExec, len(v.Ranges))
 	for i, val := range v.Ranges {
-		e.Ranges[i] = b.buildIndexRange(e, val)
+		e.Ranges[i], _ = b.buildIndexRange(e, val, idx.Meta().Columns)
 	}
-	x := b.buildFilter(e, v.FilterConditions)
+	x := b.buildFilter(e, v.AccessConditions)
+	x = b.buildFilter(x, v.FilterConditions)
 	if v.Desc {
 		x = &ReverseExec{Src: x}
 	}
 	return x
 }
 
-func (b *executorBuilder) buildIndexRange(scan *IndexScanExec, v *plan.IndexRange) *IndexRangeExec {
+// refineRange may change the IndexRange taking prefix index length into consideration
+func refineRange(v *plan.IndexRange, ics []*model.IndexColumn) bool {
+	var prefixIndex bool
+	for i, ic := range ics {
+		if ic.Length == types.UnspecifiedLength {
+			continue
+		}
+		// if index prefix length is used, change scan range
+		lowVal := &v.LowVal[i]
+		if lowVal.Kind() == types.KindBytes || lowVal.Kind() == types.KindString {
```

Comment:
The lowVal or hiVal type may not be the column type, we should check the index column's field type, which can be found in *model.Columnnfo .